### PR TITLE
kuberuntime: check the value of RunAsNonRoot when verifying

### LINKED ISSUE #46996: kubelet fails to admit containers with valid securitycontexts
**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): kubelet securitycontext

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`): 1.7.0-beta/1.8.0 (HEAD)

**Environment**:
- **Cloud provider or hardware configuration**: gce
- **OS** (e.g. from /etc/os-release): gci
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**: docker runtime (rkt seems to have a different code path, but I haven't checked it)

**What happened**: containers could not run as root when "RunAsNonRoot=false"

**What you expected to happen**: containers could run as root

**How to reproduce it** (as minimally and precisely as possible): add a securitycontext to a container, don't set privileged, do set RunAsNonRoot, use an image that runs as root. The PodSecurityPolicy admission controller will always set a securitycontext if it's not specified, and cause this to be hit more often.

**Anything else we need to know**: The bug is in the checks here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/security_context.go#L73

I have a fix in https://github.com/kubernetes/kubernetes/pull/46064 but that going in depends on an extension being granted. We may want to fix the bug regardless.

### PR DESCRIPTION
**Pull request by yujuhong**
The verification function is fixed to check the value of RunAsNonRoot,
not just the existence of it. Also adds unit tests to verify the correct
behavior.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #46996

**Release note**:

```release-note
Fix the bug where container cannot run as root when SecurityContext.RunAsNonRoot is false.
```

### CODE REVIEW

**Review by feiskyer — COMMENTED** [review]
👍 LGTM

**Comment by Q-Lee** [comment]
@liggitt this should fix the kubelet issue we discussed earlier.

**Comment by liggitt** [comment]
LGTM, would like a second from @pweil-

File: pkg/kubelet/kuberuntime/security_context_test.go [hunk 1]
```diff
@@ -46,59 +46,59 @@ func TestVerifyRunAsNonRoot(t *testing.T) {
...
+		{
+			desc: "Fail if container's RunAsUser is root and RunAsNonRoot is true",
+			sc: &v1.SecurityContext{
+				RunAsNonRoot: &runAsNonRootTrue,
+				RunAsUser:    &rootUser,
+			},
+			errStr: "container's runAsUser breaks non-root policy",
+		},
+		{
+			desc: "Fail if image's user is root and RunAsNonRoot is true",
```
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/security_context_test.go]: Add case `Pass if image's user is non-root and RunAsNonRoot is true`?

File: pkg/kubelet/kuberuntime/security_context_test.go [hunk 2]
```diff
@@ -46,59 +46,59 @@ func TestVerifyRunAsNonRoot(t *testing.T) {
 	}
 
 	err := verifyRunAsNonRoot(pod, &pod.Spec.Containers[0], int64(0))
-	assert.NoError(t, err)
+	assert.NoError(t, err, "should pass if SecurityContext is not set")
```
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/security_context_test.go]: Why not merge this case into the test table?

**Comment by yujuhong** [comment]
I pushed again just to remove the two redundant lines (48-49 in security_context_test.go) from the original test. Nothing else was changed.

File: pkg/kubelet/kuberuntime/security_context_test.go [hunk 3]
```diff
@@ -46,59 +46,59 @@ func TestVerifyRunAsNonRoot(t *testing.T) {
...
-						RunAsNonRoot: &RunAsNonRoot,
-					},
-				},
+		{
+			desc: "Fail if container's RunAsUser is root and RunAsNonRoot is true",
+			sc: &v1.SecurityContext{
+				RunAsNonRoot: &runAsNonRootTrue,
+				RunAsUser:    &rootUser,
+			},
+			errStr: "container's runAsUser breaks non-root policy",
```
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/security_context_test.go]: Hmm, not sure whether we want to check error string.

File: pkg/kubelet/kuberuntime/kuberuntime_container_test.go [hunk 4]
```diff
@@ -244,7 +244,7 @@ func TestGenerateContainerConfig(t *testing.T) {
 					Command:         []string{"testCommand"},
 					WorkingDir:      "testWorkingDir",
 					SecurityContext: &v1.SecurityContext{
-						RunAsNonRoot: &RunAsNonRoot,
+						RunAsNonRoot: &runAsNonRootTrue,
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_container_test.go]: Probably can be removed. @feiskyer 
But I don't want to turn this into a test cleanup PR. That can be done in a separate PR.

File: pkg/kubelet/kuberuntime/security_context_test.go (hunk 3 above)
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/security_context_test.go]: I didn't want to change the original test behavior too much, but remove it any way. Done.

File: pkg/kubelet/kuberuntime/security_context_test.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/security_context_test.go]: Done.

File: pkg/kubelet/kuberuntime/security_context.go [hunk 5]
```diff
@@ -72,7 +72,8 @@ func (m *kubeGenericRuntimeManager) determineEffectiveSecurityContext(pod *v1.Po
 // verifyRunAsNonRoot verifies RunAsNonRoot.
 func verifyRunAsNonRoot(pod *v1.Pod, container *v1.Container, uid int64) error {
 	effectiveSc := securitycontext.DetermineEffectiveSecurityContext(pod, container)
-	if effectiveSc == nil || effectiveSc.RunAsNonRoot == nil {
+	// If the option is not set, or if running as root is allowed, return nil.
+	if effectiveSc == nil || effectiveSc.RunAsNonRoot == nil || !*effectiveSc.RunAsNonRoot {
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/security_context.go]: This is a k8s api pod object. I don't think we have getters for them.

File: pkg/kubelet/kuberuntime/security_context_test.go (hunk 2 above)
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/security_context_test.go]: I removed this before you left the comment ;)

**Comment by liggitt** [comment]
@k8s-bot pull-kubernetes-unit test this

File: pkg/kubelet/kuberuntime/kuberuntime_container_test.go (hunk 4 above)
Comment (feiskyer) [inline pkg/kubelet/kuberuntime/kuberuntime_container_test.go]: Missing one assertion

```
	assert.Equal(t, expectedConfig, containerConfig, "generate container config for kubelet runtime v1.")
```

File: pkg/kubelet/kuberuntime/security_context_test.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/security_context_test.go]: ACK

File: pkg/kubelet/kuberuntime/security_context.go (hunk 5 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/security_context.go]: ACK. Confused initially.

**Comment by Random-Liu** [comment]
LGTM

**Comment by Q-Lee** [comment]
@k8s-bot pull-kubernetes-unit test this
#46244

**Comment by pweil-** [comment]
LGTM

**Comment by yujuhong** [comment]
I haven't squashed yet... :-|

**Comment by feiskyer** [comment]
> I haven't squashed yet... :-|

sorry, didn't notice there is another commit 😢

**Pull request merged**